Lock screen

gemdev111 opened the pull request:
Added BiometryAuthService, responsible for handling local auth and disable/enable it
Added picker for lock time period in security
Adjusted logic for lock scene  (support select time period from security)
Fixed launch screen for ipad
Welcome Scene & Launch & Lock now consistent
Added Test for LockSceneViewModel and mock for biometry auth

images:
<table>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/bfcd9b9f-ea56-4e19-94dd-3adf95546c05" style="width: 45%;" /></td>
    <td><img src="https://github.com/user-attachments/assets/260e870f-cea4-427c-adb0-b98ed7fcc30d" style="width: 45%;" /></td>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/77773c9b-b2e4-4a5b-9fba-6505dd7a3b6a" style="width: 45%;" /></td>
    <td><img src="https://github.com/user-attachments/assets/87826f35-29d8-4046-98ff-3d7c569ea884" style="width: 45%;" /></td>
  </tr>
</table>

resolve #147

gemdev111 committed:
Added BiometryService

Integrated new BiometryService, which response for handling bio auth and enable/disable it
Added LockScene & vm

gemdev111 committed:
Merge branch 'main' into lock-screen

gemdev111 committed:
Added picker for lock time period in security
Adjusted logic for lock scene to support select time period from security
Minor updates

gemdev111 committed:
adjusted lockscene view model

gemdev111 committed:
Updates Lock screen to support welcome view instead of blur effect
Authentication fixes typo, adjusted logic
Fixed launch screen for ipad
Welcome Scene & Launch & Lock now consistent
Added Test for LockSceneViewModel and mock for biometry auth

gemdev111 committed:
Merge branch 'main' into lock-screen

gemdev111 committed:
Updates localizations

gemcoder21 commented on Gem/Settings/Scenes/SecurityScene.swift:
```diff
@@ -6,65 +6,46 @@ import LocalAuthentication
-        .alert(item: $isPresentingErrorMessage) {
+        .onChange(of: model.isEnabled, onToggleEnable)
+        .alert(item: $model.isPresentingError) {
             Alert(title: Text("Transfer Error"), message: Text($0))
```
```suggestion
            Alert(title: Text("Transfer Error"), message: Text($0))
```

different title for security screen?

gemcoder21 commented on Gem/Core/Coordinator/WalletCoordinator.swift:
```diff
@@ -155,32 +152,34 @@ struct WalletCoordinator: View {
-                    Button(Localized.Common.cancel, role: .cancel) { }
-                    Button(Localized.UpdateApp.action, role: .none) {
-                        UIApplication.shared.open(PublicConstants.url(.appStore))
+                LockScreenScene(model: LockSceneViewModel()) {
```
```suggestion
                LockScreenScene(model: LockSceneViewModel()) {
```

can we move this model to the property?

gemcoder21 commented on Gem/Info.plist:
```diff
@@ -1,36 +1,31 @@
+	</array>
+	<key>ITSAppUsesNonExemptEncryption</key>
+	<false/>
+</dict>
```
`NSCameraUsageDescription` and `NSFaceIDUsageDescription` has been removed?

gemdev111 replied:
info.plist not in use for permissions. We don't need to specify them in here

Targets->Gem->Info we have the usage descriptions there 
<img width="733" alt="image" src="https://github.com/user-attachments/assets/a6ca1048-337a-4237-8184-c3b736280d8f">

gemcoder21 commented on Packages/Store/Sources/Stores/PreferencesStore.swift:
```diff
@@ -63,7 +63,10 @@ public class PreferencesStore {
-    
+
+    @UserDefault(defaults: .standard, key: "authentication_lock_option", defaultValue: 0)
+    public var authenticationLockOption: Int
```
let's move this option to keychain?

gemcoder21 commented on Packages/Keystore/Sources/Errors/BiometryAuthenticationError.swift:
```diff
@@ -0,0 +1,38 @@
+
+public enum BiometryAuthenticationError: Error {
+    case biometryUnavailable
+    case canceled
```
```suggestion
    case cancelled
```

gemcoder21 commented on Packages/Keystore/Sources/Errors/BiometryAuthenticationError.swift:
```diff
@@ -0,0 +1,38 @@
+        }
+    }
+
+    public var isAuthCancelled: Bool {
```
```suggestion
    public var isAuthenticationCancelled: Bool {
```

gemcoder21 commented on GemTests/Welcome/ViewModels/LockSceneViewModelTests.swift:
```diff
@@ -0,0 +1,397 @@
+// MARK: - Mock
+
+// TODO: - probably move to Keystore TestKip
+fileprivate class MockBiometryAuthenticationService: BiometryAuthenticable {
```
can you move this to KeystoreTestKit?

0xh3rman committed:
Merge branch 'main' into lock-screen

0xh3rman commented:
@gemdev111 I merged the `justfile` fix for you, you might need to reset SPM cache and restart Xcode

0xh3rman committed:
Fix test for Xcode 16

0xh3rman commented on Gem/Settings/Scenes/SecurityScene.swift:
```diff
@@ -6,65 +6,46 @@ import LocalAuthentication
+            Toggle(model.authenticationTitle, isOn: $model.isEnabled)
+            .toggleStyle(AppToggleStyle())
+
+            NavigationLink(value: model.lockPeriodModel.selectedOption) {
```
Should we only show this when "Enable Passcode" is enabled?

gemdev111 replied:
good point, we don't need it if disabled. Updated

0xh3rman reviewed (commented):
left some comments

0xh3rman commented on Gem/Settings/ViewModels/SecurityViewModel.swift:
```diff
@@ -3,38 +3,56 @@
+
+@Observable
+class SecurityViewModel {
+    private let service: BiometryAuthenticable
```
```suggestion
    private let service: BiometryAuthenticatable
```

0xh3rman commented on Packages/Keystore/Sources/Protocols/BiometryAuthenticable.swift:
```diff
@@ -0,0 +1,22 @@
+import Foundation
+import LocalAuthentication
+
+public protocol BiometryAuthenticable {
```
`Authenticatable` looks more suitable

0xh3rman reviewed (commented):
It seems you can input any passcode on Simulator to disable Passcode, is it expected?

0xh3rman commented:
Should we block WalletConnect or payment deeplink request? how to reproduce:
1. Enable Passcode (Immediate)
2. Open Uniswap and select Gem in WalletConnect
3. The screen below will show on top of lock screen

![Simulator Screenshot - iPhone 16 Pro - 2024-09-19 at 14 30 17](https://github.com/user-attachments/assets/636c03d0-2344-4edf-8260-a4459987084d)

gemdev111 commented:
> It seems you can input any passcode on Simulator to disable Passcode, is it expected?

yeah, on a simulator, if you input any symbol in "apple auth view"- you got success authentication.

gemdev111 committed:
typo & naming updates

gemdev111 committed:
Merge remote-tracking branch 'refs/remotes/origin/lock-screen'

gemdev111 committed:
show lock perioid picker only if enabled, error title updated

gemdev111 committed:
LockPeriod moved to keysore & keychain
minor adjustemnets

gemdev111 committed:
Merge branch 'main' into lock-screen

gemdev111 committed:
Added WalletConnectManager responsible for handling wc-requests and transform existed requests to pending, if device in locked state

gemdev111 committed:
Updates lock-scene

Added LockWindowManager into app module, it's responsible for showing/hiding lock-window
Removed lock model from coordinator
Wallet list item fixed crash with force unwrapping, when removing last wallet
Added cleanup for biometry authentication, when removing last wallet
Added tests for LockWindowManager

gemdev111 committed:
Merge branch 'main' into lock-screen

gemcoder21 committed:
Remove #if RELEASE

gemcoder21 committed:
Move GemDev.plist to /Gem

gemcoder21 committed:
Update project.pbxproj

gemcoder21 committed:
Update project.pbxproj

gemdev111 committed:
removed animation show lock

gemdev111 committed:
Merge remote-tracking branch 'refs/remotes/origin/lock-screen'

gemdev111 committed:
Added ConfirmTransferViewModel, adjusted lock images based on type
Clenup temporary commented LocalKeystorePassword
